tuf/ngclient/_internal/metadata_bundle.py

```diff
@@ -0,0 +1,461 @@
+ * Progress through Specification update process should be documented
+   (not sure yet how: maybe a spec_logger that logs specification events?)
+"""
+
+import logging
+from collections import abc
+from datetime import datetime
+from typing import Dict, Iterator, Optional
+
+from securesystemslib import hash as sslib_hash
+from securesystemslib import keys as sslib_keys
+
+from tuf import exceptions
+from tuf.api.metadata import Metadata, Root, Targets
+from tuf.api.serialization import DeserializationError
+
+logger = logging.getLogger(__name__)
+
+# This is a placeholder until ...
+# TODO issue 1306: implement this in Metadata API
+def verify_with_threshold(
+    delegator: Metadata, role_name: str, unverified: Metadata
+) -> bool:
+    """Verify 'unverified' with keys and threshold defined in delegator"""
+    role = None
+    keys = {}
+    if isinstance(delegator.signed, Root):
+        keys = delegator.signed.keys
+        role = delegator.signed.roles.get(role_name)
+    elif isinstance(delegator.signed, Targets):
+        if delegator.signed.delegations:
+            keys = delegator.signed.delegations.keys
+            # role names are unique: first match is enough
+            roles = delegator.signed.delegations.roles
+            role = next((r for r in roles if r.name == role_name), None)
+    else:
+        raise ValueError("Call is valid only on delegator metadata")
+
+    if role is None:
+        raise ValueError(f"Delegated role {role_name} not found")
+
+    # verify that delegate is signed by correct threshold of unique keys
+    unique_keys = set()
+    for keyid in role.keyids:
+        key_dict = keys[keyid].to_dict()
+        key, dummy = sslib_keys.format_metadata_to_key(key_dict)
+
+        try:
+            if unverified.verify(key):
+                unique_keys.add(key["keyval"]["public"])
+        except Exception as e:  # pylint: disable=broad-except
+            # TODO specify the Exceptions (see issue #1351)
+            logger.info("verify failed: %s", e)
+
+    return len(unique_keys) >= role.threshold
+
+
+class MetadataBundle(abc.Mapping):
+    """Internal class to keep track of valid metadata in Updater
+
+    MetadataBundle ensures that the collection of metadata in the bundle is
+    valid. It provides easy ways to update the metadata with the caller making
+    decisions on what is updated.
+    """
+
+    def __init__(self, root_data: bytes):
+        """Initialize bundle by loading trusted root metadata
+
+        Args:
+            root_data: Trusted root metadata as bytes. Note that this metadata
+                will only be verified by itself: it is the source of trust for
+                all metadata in the bundle.
+
+        Raises:
+            RepositoryError: Metadata failed to load or verify. The actual
+                error type and content will contain more details.
+        """
+        self._bundle = {}  # type: Dict[str: Metadata]
+        self.reference_time = datetime.utcnow()
```

Comment:
hmm, maybe? 

Why is there a grace period? I can think of two potential cases
* clients with really bad clocks 
* buggy repositories that do publish new metadata but do it after old one has expired

This still sounds like something that isn't part of this PR (ngclient is already better than current code since we at least have a reference time). We should file an issue if this is relevant

Reply:
There's an issue filed against the spec to [add the notion of slack time](https://github.com/theupdateframework/specification/issues/137). I propose we hold off on adding this feature to the reference implementation until it is part of the spec.